File: packages/python_ffi/example/python-modules/structs.py
class Coordinate:
    def __init__(self, latitude: float, longitude: float) -> 'Coordinate':
        self.latitude = latitude
        self.longitude = longitude


class Place:
    def __init__(self, name: str, coordinate: Coordinate) -> 'Place':
        self.name = name
        self.coordinate = coordinate


def hello_world() -> str:
    return "Hello World"


def reverse(str: str, length: int) -> str:
    return str[::-1]


def create_coordinate(latitude: float, longitude: float) -> Coordinate:
    return Coordinate(latitude, longitude)


def create_place(name: str, latitude: float, longitude: float) -> Place:
    coordinate: Coordinate = Coordinate(latitude, longitude)
    return Place(name, coordinate)


def distance(c1: Coordinate, c2: Coordinate) -> float:
    xd: float = c2.latitude - c1.latitude
    yd: float = c2.longitude - c1.longitude
    return (xd * xd + yd * yd) ** 0.5

File: packages/python_ffi/example/lib/python-modules/structs.dart
import "package:python_ffi/python_ffi.dart";

class Coordinate extends PythonClass {
  factory Coordinate(double latitude, double longitude) {
    final Coordinate coordinate = PythonFfi.instance.importClass(
      "structs",
      "Coordinate",
      Coordinate.from,
      <Object?>[latitude, longitude],
    );
    return coordinate;
  }

  Coordinate.from(super.pythonClass) : super.from();

  @override
  StructsModule get module => StructsModule.import();

  double get latitude => (getAttribute("latitude")! as num).toDouble();

  double get longitude => (getAttribute("longitude")! as num).toDouble();

  @override
  String toString() => "Coordinate(latitude: $latitude, longitude: $longitude)";
}

class StructsModule extends PythonModule {
  StructsModule.from(super.pythonModule) : super.from();

  static StructsModule import() => PythonFfi.instance.importModule(
        "structs",
        StructsModule.from,
      );

  String hello_world() => getFunction("hello_world").call(<Object?>[]);

  String reverse(String str, int length) =>
      getFunction("reverse").call(<Object?>[str, length]);

  Coordinate create_coordinate(int a, int b) => Coordinate.from(
        getFunction("create_coordinate").call(
          <Object?>[a, b],
        ),
      );
}
